File: ios/LookSeeProto/UI Files/VariableContainer.swift
//
//  VariableContainer.swift
//  LookSeeProto
//
//  Created by Christian Barbara on 4/12/26.
//

import SwiftUI
import Combine

@MainActor
final class VariableContainer: ObservableObject {
    static let shared = VariableContainer()

    @Published var infoView: Bool = false
    @Published var bboxCounter: Int = 0

    @Published var landmarkName: String = "Not available"
    @Published var landmarkConfidence: Float = 0.00
    @Published var landmarkCategory: String = "Not available"
    @Published var landmarkDescription: String = "No description is available for this landmark."
    @Published var landmarkURL: String = ""
    @Published var landmarkWebsiteUrl: String = ""

    @Published var landmarkId: String = ""
    @Published var landmarkClassIndex: Int?
    @Published var landmarkClusterId: Int?
    @Published var landmarkTrainingRunId: String = ""
    @Published var landmarkDatasetClassName: String = ""

    @Published var promoName: String = "No active promotion"
    @Published var promoDescription: String = ""
    @Published var promoImageUrl: String = ""
    
    @Published var merchantName: String = ""
    @Published var merchantBio: String = ""
    @Published var merchantPhone: String = ""
    @Published var merchantWebsite: String = ""
    @Published var merchantAddress: String = ""
    @Published var merchantLogoUrl: String = ""
    
    // 🚀 NEW: Map Routing & Reporting Fields
    @Published var isMapPin: Bool = false
    @Published var mapLatitude: Double?
    @Published var mapLongitude: Double?
    @Published var reportedOwnerId: String?

    private init() {
        resetLandmarkDisplay()
    }

    func presentLandmark(
        _ entry: LandmarkManifestEntry,
        clusterId: Int,
        trainingRunId: String,
        detectionConfidence: Float
    ) {
        resetLandmarkDisplay()
        
        landmarkId = entry.landmarkId
        landmarkClassIndex = entry.classIndex
        landmarkClusterId = clusterId
        landmarkTrainingRunId = trainingRunId
        landmarkDatasetClassName = entry.datasetClassName

        landmarkName = entry.label

        let trimmedDescription = entry.shortDescription
            .trimmingCharacters(in: .whitespacesAndNewlines)

        landmarkDescription = trimmedDescription.isEmpty
            ? "No description is available for this landmark."
            : trimmedDescription

        landmarkCategory = entry.datasetClassName
            .replacingOccurrences(of: "_", with: " ")

        let clampedConfidence = min(max(detectionConfidence, 0), 1)
        landmarkConfidence = clampedConfidence * 100
        
        let cacheKey = "cached_merchant_\(entry.landmarkId)"
        if let cachedData = UserDefaults.standard.dictionary(forKey: cacheKey) {
            merchantName = cachedData["merchantName"] as? String ?? ""
            merchantBio = cachedData["merchantBio"] as? String ?? ""
            merchantPhone = cachedData["merchantPhone"] as? String ?? ""
            merchantWebsite = cachedData["merchantWebsite"] as? String ?? ""
            merchantAddress = cachedData["merchantAddress"] as? String ?? ""
            merchantLogoUrl = cachedData["merchantLogoUrl"] as? String ?? ""
        }

        infoView = true
    }

    // 🚀 NEW: Added promoName, promoDescription, and Merchant Cache lookup!
    func presentMapLandmark(
        id: String,
        name: String,
        description: String,
        latitude: Double,
        longitude: Double,
        promotionEnabled: Bool,
        promotion: String?,
        ownerId: String?,
        websiteUrl: String?,
        promoName: String?,         // <-- NEW
        promoDescription: String?,  // <-- NEW
        promoImageUrl: String?,
        merchantName: String?,
        merchantBio: String?,
        merchantPhone: String?,
        merchantAddress: String?,
        merchantLogoUrl: String?
    ) {
        resetLandmarkDisplay()
        
        self.landmarkId = id
        self.landmarkName = name
        
        let trimmedDesc = description.trimmingCharacters(in: .whitespacesAndNewlines)
        self.landmarkDescription = trimmedDesc.isEmpty ? "No description is available for this landmark." : trimmedDesc
        
        self.isMapPin = true
        self.mapLatitude = latitude
        self.mapLongitude = longitude
        self.reportedOwnerId = ownerId
        
        self.landmarkWebsiteUrl = websiteUrl ?? ""
        
        if promotionEnabled {
            self.promoName = promoName ?? promotion ?? "Special Promotion Available!"
            self.promoDescription = promoDescription ?? ""
            self.promoImageUrl = promoImageUrl ?? ""
        }
        
        // 🚀 THE FIX: Pull the Merchant Info from the exact same cache the Scanner uses!
        let cacheKey = "cached_merchant_\(id)"
        if let cachedData = UserDefaults.standard.dictionary(forKey: cacheKey) {
            self.merchantName = cachedData["merchantName"] as? String ?? merchantName ?? ""
            self.merchantBio = cachedData["merchantBio"] as? String ?? merchantBio ?? ""
            self.merchantPhone = cachedData["merchantPhone"] as? String ?? merchantPhone ?? ""
            self.merchantWebsite = cachedData["merchantWebsite"] as? String ?? merchantWebsite ?? ""
            self.merchantAddress = cachedData["merchantAddress"] as? String ?? merchantAddress ?? ""
            self.merchantLogoUrl = cachedData["merchantLogoUrl"] as? String ?? merchantLogoUrl ?? ""
        } else {
            self.merchantName = merchantName ?? ""
            self.merchantBio = merchantBio ?? ""
            self.merchantPhone = merchantPhone ?? ""
            self.merchantWebsite = merchantWebsite ?? ""
            self.merchantAddress = merchantAddress ?? ""
            self.merchantLogoUrl = merchantLogoUrl ?? ""
        }
        
        self.infoView = true
    }

    func dismissLandmark() {
        infoView = false
        landmarkURL = ""
    }

    func resetLandmarkDisplay() {
        infoView = false
        bboxCounter = 0

        landmarkName = ""
        landmarkConfidence = 0.00
        landmarkCategory = "Not available"
        landmarkDescription = "No description is available for this landmark."
        landmarkURL = ""
        landmarkWebsiteUrl = ""

        landmarkId = ""
        landmarkClassIndex = nil
        landmarkClusterId = nil
        landmarkTrainingRunId = ""
        landmarkDatasetClassName = ""

        promoName = "No active promotion"
        promoDescription = ""
        promoImageUrl = ""
        
        merchantName = ""
        merchantBio = ""
        merchantPhone = ""
        merchantWebsite = ""
        merchantAddress = ""
        merchantLogoUrl = ""
        
        isMapPin = false
        mapLatitude = nil
        mapLongitude = nil
        reportedOwnerId = nil
    }

    func getlandmarkName() -> String {
        landmarkName
    }
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/uifiles/VariableContainer.kt
package looksee.angelll.com.uifiles // Added your requested package name

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * Shared display state used by the landmark information popup.
 *
 * The manifest-based detection flow resolves a `LandmarkManifestEntry`, then
 * calls `presentLandmark(...)` to populate the same fields already consumed by
 * `LandmarkInfo` and `PopUp`.
 */
class VariableContainer private constructor() {

    companion object {
        val shared = VariableContainer()

        // Hold a reference to the app context for SharedPreferences access
        private var appContext: Context? = null

        fun initialize(context: Context) {
            appContext = context.applicationContext
        }
    }

    var infoView by mutableStateOf(false)
    var bboxCounter by mutableStateOf(0)

    var landmarkName by mutableStateOf("Not available")
    var landmarkConfidence by mutableStateOf(0.00f)
    var landmarkCategory by mutableStateOf("Not available")
    var landmarkDescription by mutableStateOf("No description is available for this landmark.")
    var landmarkURL by mutableStateOf("")
    var landmarkWebsiteUrl by mutableStateOf("")

    // Manifest/debugging identity. These are useful when confirming that a
    // popup was resolved from the same cluster release as the active model.
    var landmarkId by mutableStateOf("")
    var landmarkClassIndex by mutableStateOf<Int?>(null)
    var landmarkClusterId by mutableStateOf<Int?>(null)
    var landmarkTrainingRunId by mutableStateOf("")
    var landmarkDatasetClassName by mutableStateOf("")

    var promoName by mutableStateOf("No active promotion")
    var promoDescription by mutableStateOf("")
    var promoImageUrl by mutableStateOf("")

    // 🚀 NEW: Enterprise Merchant Profile Fields
    var merchantName by mutableStateOf("")
    var merchantBio by mutableStateOf("")
    var merchantPhone by mutableStateOf("")
    var merchantWebsite by mutableStateOf("")
    var merchantAddress by mutableStateOf("")
    var merchantLogoUrl by mutableStateOf("")

    init {
        resetLandmarkDisplay()
    }

    /**
     * Populates and opens the popup using information resolved from a local
     * cluster landmark manifest.
     *
     * @param entry The manifest entry associated with the detected class index.
     * @param clusterId Cluster whose model produced the detection.
     * @param trainingRunId Immutable release identifier paired with the model.
     * @param detectionConfidence Raw model confidence in the range 0...1.
     */
    fun presentLandmark(
        entry: LandmarkManifestEntry,
        clusterId: Int,
        trainingRunId: String,
        detectionConfidence: Float
    ) {
        landmarkId = entry.landmarkId
        landmarkClassIndex = entry.classIndex
        landmarkClusterId = clusterId
        landmarkTrainingRunId = trainingRunId
        landmarkDatasetClassName = entry.datasetClassName

        landmarkName = entry.label

        val trimmedDescription = entry.shortDescription.trim()

        landmarkDescription = if (trimmedDescription.isEmpty()) {
            "No description is available for this landmark."
        } else {
            trimmedDescription
        }

        // Keep the existing UI field populated without exposing the S3 folder
        // formatting directly to the user.
        landmarkCategory = entry.datasetClassName.replace("_", " ")

        val clampedConfidence = detectionConfidence.coerceIn(0f, 1f)
        landmarkConfidence = clampedConfidence * 100

        // Live website/promotion data is fetched from the backend after the
        // popup opens. Reset these so a previous landmark cannot leak into the
        // newly displayed popup.
        landmarkWebsiteUrl = ""
        promoName = "No active promotion"
        promoDescription = ""
        promoImageUrl = ""
        landmarkURL = ""

        // 🚀 INSTANT CACHE LOAD: Load cached merchant details immediately to eliminate delay
        appContext?.let { context ->
            val prefs = context.getSharedPreferences("LookSeeMerchantCache", Context.MODE_PRIVATE)
            val cacheKeyPrefix = "cached_merchant_${entry.landmarkId}_"

            merchantName = prefs.getString("${cacheKeyPrefix}merchantName", "") ?: ""
            merchantBio = prefs.getString("${cacheKeyPrefix}merchantBio", "") ?: ""
            merchantPhone = prefs.getString("${cacheKeyPrefix}merchantPhone", "") ?: ""
            merchantWebsite = prefs.getString("${cacheKeyPrefix}merchantWebsite", "") ?: ""
            merchantAddress = prefs.getString("${cacheKeyPrefix}merchantAddress", "") ?: ""
            merchantLogoUrl = prefs.getString("${cacheKeyPrefix}merchantLogoUrl", "") ?: ""
        } ?: run {
            merchantName = ""
            merchantBio = ""
            merchantPhone = ""
            merchantWebsite = ""
            merchantAddress = ""
            merchantLogoUrl = ""
        }

        infoView = true

        println("""
        ✅ [Local Manifest Popup] Presenting landmark
            clusterId: $clusterId
            trainingRunId: $trainingRunId
            classIndex: ${entry.classIndex}
            landmarkId: ${entry.landmarkId}
            label: ${entry.label}
            confidence: ${String.format("%.1f", landmarkConfidence)}%
        """.trimIndent())
    }

    /**
     * Closes the popup while retaining the most recently displayed landmark
     * values for diagnostics.
     */
    fun dismissLandmark() {
        infoView = false
        landmarkURL = ""
    }

    /**
     * Clears all landmark-specific state. This can be used when changing
     * active cluster releases or resetting the scan screen.
     */
    fun resetLandmarkDisplay() {
        infoView = false
        bboxCounter = 0

        landmarkName = ""
        landmarkConfidence = 0.00f
        landmarkCategory = "Not available"
        landmarkDescription = "No description is available for this landmark."
        landmarkURL = ""
        landmarkWebsiteUrl = ""

        landmarkId = ""
        landmarkClassIndex = null
        landmarkClusterId = null
        landmarkTrainingRunId = ""
        landmarkDatasetClassName = ""

        promoName = "No active promotion"
        promoDescription = ""
        promoImageUrl = ""

        merchantName = ""
        merchantBio = ""
        merchantPhone = ""
        merchantWebsite = ""
        merchantAddress = ""
        merchantLogoUrl = ""
    }
}

// Stub for LandmarkManifestEntry to prevent compiler errors
data class LandmarkManifestEntry(
    val landmarkId: String,
    val classIndex: Int,
    val datasetClassName: String,
    val label: String,
    val shortDescription: String
)
